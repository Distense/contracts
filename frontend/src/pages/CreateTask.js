import React, { Component } from 'react'
import CodeMirror from 'react-codemirror2'
import 'codemirror/lib/codemirror.css'
import 'codemirror/mode/markdown/markdown'
import { Dropdown } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router'
import ReactMarkdown from 'react-markdown'
import slug from 'slug'

import { getPendingTask } from '../reducers/tasks'
import { createTask } from '../actions'
import Head from '../components/common/Head'
import Layout from '../components/Layout'
import { tagsOptions, specPlaceholder } from '../shared'

const taskUrl = ({ title, _id }) => `/tasks/${slug(title)}/${_id}`


class CreateTask extends Component {
  constructor(props) {
    super(props)
    this.state = {
      title: '',
      tags: [],
      spec: specPlaceholder
    }

    this.onChangeTags = this.onChangeTags.bind(this)
  }

  onChangeTitle = ({ target: { value } }) => {
    if (value.length <= 50)
      this.setState({ title: value })
  }

  onChangeTags(e, data) {
    const tags = data.value
    if (tags.length < 6)
      this.setState({ tags })
  }

  onChangeSpec = (editor, metadata, value) => {
    this.setState({ spec: value })
  }

  onSubmit = async (e) => {
    e.preventDefault()
    const { title, tags, spec } = this.state
    this.props.createTask({ title, tags, spec })
  }

  render() {
    const { pendingTask } = this.props
    const {
      title,
      tags,
      spec
    } = this.state

    if (pendingTask) {
      //  TODO probably redirect to /tasks
      //  This does not allow navigating to /tasks
      //  then navigating to CreateTask while !!pendingTask
      return <Redirect to={taskUrl(pendingTask)}/>
    }

    return (
      <Layout>
        <Head title='Create Task'/>
        <div className='ui one column grid'>
          <form className='ui form' onSubmit={this.onSubmit}>
            <h1 className='ui header'>Create Task</h1>
            <div className='field'>
              <input
                type='text'
                placeholder='Title'
                onChange={this.onChangeTitle}
                className=''
                name='title'
                value={title}
              />
            </div>
            <div className='field'>
              <Dropdown
                // compact
                fluid
                multiple
                onChange={this.onChangeTags}
                options={tagsOptions}
                placeholder='Tags'
                search
                selection
                scrolling
                value={tags}
              />
            </div>
            <div className='fields fields-margin-fix'>
              <div className='eight wide field'>
                <CodeMirror value={spec}
                  options={{
                    cursorBlinkRate: 650,
                    lineNumbers: true,
                    lineWrapping: true,
                    mode: 'markdown',
                    scrollbarStyle: null,
                    tabSize: 2
                  }}
                  onValueChange={this.onChangeSpec}
                />
              </div>
              <div className='eight wide field react-markdown'>
                <ReactMarkdown source={spec} />
              </div>
            </div>
            <button className='ui large green button' type='submit'>
              Submit
            </button>
          </form>
        </div>

        { /*language=CSS*/ }
        <style global jsx>{`

          .react-codemirror2, .react-markdown {
            border: 1px solid rgba(34,36,38,.15);
            border-radius: .28571429rem;
          }

          .react-codemirror2 {
            border-top-right-radius: 0;
            border-bottom-right-radius: 0;
            margin-right: -.5rem;
          }

          .react-markdown {
            border-top-left-radius: 0;
            border-bottom-left-radius: 0;
            padding: 1rem;
          }

          .fields-margin-fix {
            margin: 0 -.5em 1em !important;
          }

          .CodeMirror {
            height: 36rem;
          }

          .CodeMirror-scroll {
            padding-bottom: 0;
          }

          .center {
            text-align: center;
          }

          .underlined {
            border-bottom: 1px solid #ccc;
          }

          .CodeMirror-cursor {
            border-left: .5rem solid #ccc;
          }

        `}</style>
      </Layout>
    )
  }
}
const mapStateToProps = (state) => ({
  pendingTask: getPendingTask(state)
})

const mapDispatchToProps = dispatch => ({
  createTask: task => dispatch(createTask(task))
})

export default connect(mapStateToProps, mapDispatchToProps)(CreateTask)