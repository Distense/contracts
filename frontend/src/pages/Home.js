import React, { Component } from 'react'
import Icon from 'react-fontawesome'
import { Motion, spring } from 'react-motion'
import windowSize from 'react-window-size'

import scrollPosition from '../components/scrollPosition'
import Head from '../components/common/Head'
import Sky from '../components/Sky'

const heroTop = (height, top) => Math.max(0, (height / 3) * (1 - (top / height)))

class Home extends Component {
  constructor(props) {
    super(props)

    this.state = {
      email: '',
      emailFocused: false
    }
  }

  onFocusEmail = (e) => {
    this.setState({ emailFocused: true })
  }

  onBlurEmail = (e) => {
    this.setState({ emailFocused: false })
  }

  onChangeEmail = ({ target: { value: email } }) => {
    this.setState({ email })
  }

  onSubmitEmail = (e) => {
    e.preventDefault()

    this.setState({
      emailSubmitSuccess: true
    })

    fetch('https://xe6au48aog.execute-api.us-west-2.amazonaws.com/prod/mailchimpLambda', {
      method: 'POST',
      body: JSON.stringify({ email: this.state.email })
    })
  }

  render() {
    const { windowWidth, windowHeight } = this.props
    const { email, emailFocused, emailSubmitSuccess } = this.state
    const percentOfDay = emailSubmitSuccess ? 0.9 : (email || emailFocused ? 0.98 : 1)

    return (
      <div className='container'>
        <Head />

        <div className='sky'>
          <Motion style={{ percentOfDay: spring(percentOfDay) }}>
            {({ percentOfDay }) =>
              <Sky
                width={windowWidth}
                height={windowHeight}
                percentOfDay={percentOfDay}
              />
            }
          </Motion>
        </div>

        <header className='hero' style={{ top: `${heroTop(windowHeight, scrollPosition)}px` }}>
          <div className='header-inner max-width'>
            <div>
              <h1>Distense</h1>
              <h2>A decentralized, for-profit code cooperative</h2>
              {emailSubmitSuccess ?
                <div className='email-form-success'>
                  Thanks, we'll update you soon!
                </div>
                :
                <form className='email-form' onSubmit={this.onSubmitEmail}>
                  <input
                    ref={i => this.input = i}
                    type='email'
                    placeholder='Email me updates'
                    value={email}
                    onFocus={this.onFocusEmail}
                    onBlur={this.onBlurEmail}
                    onChange={this.onChangeEmail}
                  />
                  <button type='submit' className={email && 'show'} disabled={!email}>
                    <Icon name={emailSubmitSuccess ? 'check' : 'chevron-right'}/>
                  </button>
                </form>
              }
              <div className="overview">
                <a target="_blank" href="../public/Distense-Overview-8-17-17.pdf" download>
                  <h2 className="underlined">One Page Overview</h2>
                </a>
              </div>
            </div>
          </div>
        </header>

        <div className='content'>
          <header style={{ top: `${heroTop(windowHeight, scrollPosition)}px` }}>
            <div className='header-inner max-width'>
              <h1 className='logo'>D</h1>
              <h1 style={{ visibility: 'hidden' }}>D</h1>
            </div>
          </header>

          <div className='max-width'>

          </div>
        </div>

        <style jsx>{`
          .container {
            padding-top: 100vh;
          }

          .show {
            opacity: 1;
          }

          .underlined {
            text-decoration: underline;
          }

          .overview {
            margin-top: 2rem;
          }

          .max-width {
            width: 100%;
            max-width: 80rem;
            margin: 0 auto;
            padding: 0 2rem;
          }

          header {
            position: fixed;
            top: 0;
            left: 0;
            width: 100vw;
          }

          .header-inner {
            display: flex;
            flex-direction: row;
            justify-content: space-between;
            align-items: center;
          }

          .hero {
            top: 30vh;
          }

          .email-form, .email-form-success {
            margin-top: 3rem;
          }

          .email-form-success {
            font-size: 1.2rem;
            color: #f5eec5;
            padding: 1rem 0;
          }

          nav {
            display: flex;
            flex-direction: row;
          }

          nav a {
            color: white;
          }

          .content nav a {
            color: black;
          }

          .logo {
            position: absolute;
          }

          .content {
            display: none;
            position: absolute;
            background: white;
            width: 100vw;
            height: 200vh;
            padding: 1rem;
            overflow: hidden;
            clip: rect(auto, auto, auto, auto);
          }

          .sky {
            position: fixed;
            top: 0;
            left: 0;
            width: 100vw;
            height: 100vh;
            background: #151E2A;
          }

          h1 {
            font-size: 3rem;
          }

          h2 {
            font-size: 1.5rem;
          }

          .hero h1 {
            color: white;
          }

          .hero h2 {
            color: rgba(255, 255, 255, 0.4)
          }

          form {
            position: relative;
            display: flex;
          }

          input {
            flex: 1;
            font-size: 1.2rem;
            padding: 1rem 0;
            background: none;
            border: none;
            border-bottom: 1px solid rgba(255, 255, 255, 0.4);
            color: #fff;
          }

          input:hover {
            border-bottom-color: rgba(255, 255, 255, 0.7);
          }

          input:focus {
            border-bottom-color: #fff;
          }

          input::placeholder {
            color: rgba(255, 255, 255, 0.4);
          }

          button {
            position: absolute;
            right: 0;
            font-size: 1.5rem;
            color: rgba(255, 255, 255, 0.4);
            height: 100%;
            opacity: 0;
          }

          button:hover {
            color: #fff;
          }
        `}</style>
      </div>
    )
  }
}

export default windowSize(scrollPosition(Home))
