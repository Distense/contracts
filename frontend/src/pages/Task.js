import React, { Component } from 'react'
import { connect } from 'react-redux'

import { fetchTask } from '../actions'
import { getTask } from '../reducers/tasks'

import Head from '../components/common/Head'
import Layout from '../components/Layout'

class Task extends Component {
  componentWillMount() {
    const { fetchTask, match: { params: { id }}} = this.props
    fetchTask(id)
  }

  render() {
    const { task } = this.props

    return (
      <Layout>
        <Head title="Task"/>
        <div className="task">
          {JSON.stringify(task)}
        </div>
      </Layout>
    )
  }
}

const mapStateToProps = (state, { match: { params: { id }}}) => ({
  task: getTask(state, id)
})

const mapDispatchToProps = dispatch => ({
  fetchTask: id => dispatch(fetchTask(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(Task)
