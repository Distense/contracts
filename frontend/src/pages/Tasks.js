import React, { Component } from 'react'
import { connect } from 'react-redux'

import { fetchTasks } from '../actions'
import { getAllTasks } from '../reducers/tasks'

import Head from '../components/common/Head'
import Layout from '../components/Layout'

const Task = ({ task }) => (
  <div>
    {JSON.stringify(task)}
  </div>
)

class Tasks extends Component {
  componentWillMount() {
    this.props.fetchTasks()
  }

  render() {
    const { tasks } = this.props
    return (
      <Layout>
        <Head title="Available Tasks"/>
        <div className="tasks-container">
          <h2>Available Tasks</h2>
          <div className="tasks-list">
            {tasks.length ?
              tasks.map(task => <Task key={task._id} task={task} />
              ) : 'No Tasks Available'
            }
          </div>
        </div>
      </Layout>
    )
  }
}

const mapStateToProps = state => ({
  tasks: getAllTasks(state)
})

const mapDispatchToProps = dispatch => ({
  fetchTasks: () => dispatch(fetchTasks())
})

export default connect(mapStateToProps, mapDispatchToProps)(Tasks)