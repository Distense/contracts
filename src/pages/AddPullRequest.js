import React, { Component } from 'react'
import { connect } from 'react-redux'
import {
  Button,
  Divider,
  Form,
  Grid,
  Input,
  Header,
  List,
  Message
} from 'semantic-ui-react'
import { Link, Redirect } from 'react-router-dom'

import { fetchTask, selectTask } from '../actions/tasks'
import { getSelectedTask } from '../reducers/tasks'
import { addPullRequest } from '../actions/pullRequests'

import Head from '../components/common/Head'

export class AddPullRequest extends Component {
  constructor(props) {
    super(props)
    this.state = {
      taskId: '',
      prNum: '',
      redirect: false
    }
    this.onSubmit = this.onSubmit.bind(this)
  }

  componentDidMount() {
    //  If taskId param exists prefill the input
    const taskId = this.props.match.params.id
    if (taskId) this.selectTask(taskId)
  }

  onChangeTaskId = ({ target: { value } }) => {
    this.selectTask(value)
    // Don't fetch the task for previewing until we've at least got enough bytes to be a complete ipfs hash
  }

  selectTask(taskId) {
    this.setState({ taskId: taskId })
    this.props.selectTask(taskId)
  }

  onChangePRNum = ({ target: { value } }) => {
    this.setState({ prNum: value })
  }

  onSubmit = async e => {
    e.preventDefault()
    const { taskId, prNum } = this.state
    if (taskId && prNum) {
      this.props.createPullRequest({ taskId, prNum })
      this.setState({
        redirect: true
      })
    }
  }

  render() {
    const { task } = this.props
    const { taskId, prNum, redirect } = this.state

    if (redirect) {
      return <Redirect to="/pullrequests" />
    }

    return (
      <div>
        <Head title="Add PullRequest" />
        <Grid columns={1}>
          <Grid.Row width={4} columns={1}>
            <Form onSubmit={this.onSubmit}>
              <Header as="h1">Submit Pull Request</Header>
              <Form.Field required>
                <Input
                  label="Pull Request Number"
                  type="text"
                  placeholder=""
                  onChange={this.onChangePRNum}
                  name="url"
                  value={prNum}
                />
              </Form.Field>
              <Form.Field required>
                <Input
                  label="Task Id"
                  name="id"
                  onChange={this.onChangeTaskId}
                  placeholder="hash"
                  type="text"
                  value={taskId}
                />
              </Form.Field>
              <Button size="large" color="green" type="submit">
                Submit
              </Button>
            </Form>
          </Grid.Row>
          <Grid.Row>
            <Message>
              <Message.Header>Pull Request number</Message.Header>
              <List bulleted>
                <List.Item>First submit a pull request on Github</List.Item>
                <List.Item>
                  Enter the number found in your PR URL in the top box of this
                  web page
                </List.Item>
                <List.Item>
                  The number is the last part of your pull request URL. Example:
                  https://github.com/Distense/distense-ui/pulls/<b>321</b>
                </List.Item>
                <List.Item>Enter just the number: 123</List.Item>
              </List>
            </Message>
            <Message>
              <Message.Header>Task ID</Message.Header>
              <List bulleted>
                <List.Item>
                  Find the Task ID of the task your pull request completes
                </List.Item>
                <List.Item>
                  The Task ID can be found by clicking the corresponding task
                  found in the&nbsp;
                  <Link to="/tasks">tasks list</Link>
                  &nbsp;page.
                </List.Item>
                <List.Item>
                  If you know the URL of your task, the Task ID is the last part
                  and will look like: 1515652690755a0f3f4b21c0
                </List.Item>
              </List>
            </Message>
          </Grid.Row>
        </Grid>
        {task && (
          // Preview task info so submitter can verify they're submitting for
          // the right task
          <div>
            <Divider section />
            <Header as="h3">Pull Request Task Verification</Header>
            <Message>
              <Message.Header>Title: {task.title}</Message.Header>
            </Message>
          </div>
        )}
      </div>
    )
  }
}

const mapStateToProps = state => ({
  task: getSelectedTask(state)
})

const mapDispatchToProps = dispatch => ({
  fetchTask: id => dispatch(fetchTask(id)),
  createPullRequest: pr => dispatch(addPullRequest(pr)),
  selectTask: taskId => dispatch(selectTask(taskId))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddPullRequest)
