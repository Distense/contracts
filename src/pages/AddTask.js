import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import {
  Button,
  Dropdown,
  Input,
  Form,
  Grid,
  List,
  Message
} from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router'

import { addTask } from '../actions/tasks'
import { getPendingTask } from '../reducers/tasks'
import { NUM_DID_REQUIRED_TO_ADD_TASK_PARAMETER_TITLE } from '../constants/parameters/parameterTitles'

import Head from '../components/common/Head'
import PageTitling from '../components/common/PageTitling'
import { tagsOptions } from '../tagsOptions'
import { getParameterValueByTitle } from '../reducers/parameters'

export class AddTask extends Component {
  constructor(props) {
    super(props)
    this.state = {
      title: '',
      tagsString: '',
      issueNum: '',
      numDIDRequiredToAddTask: this.props.numDIDRequiredToAddTask || 120,
      repo: '',
      redirect: false
    }

    this.onChangeRepo = this.onChangeRepo.bind(this)
    this.onChangeTags = this.onChangeTags.bind(this)
  }

  componentDidMount() {
    this.someTimeout = setTimeout(() => {
      this.setState({
        numDIDRequiredToAddTask: this.props.numDIDRequiredToAddTask
      })
    }, 2000)
  }

  onChangeTitle = ({ target: { value } }) => {
    //  make sure not to have any slashes for URLs
    value = value.replace('/', '')
    if (value.length <= 70) {
      this.setState({ title: value })
    }
  }

  onChangeIssueNum = ({ target: { value } }) => {
    this.setState({ issueNum: value })
  }

  onChangeTags(e, data) {
    const tags = data.value
    if (tags.length < 3) {
      this.setState({ tags })
      let tagsString = ''
      tags.forEach((tag, i) => {
        if (i === 0) tagsString = tag
        else tagsString += ':' + tag
      })
      this.setState({ tagsString })
    }
  }

  onChangeRepo(e, data) {
    this.setState({
      repo: data.value
    })
  }

  onSubmit = async e => {
    e.preventDefault()

    const { title, tagsString, issueNum, repo } = this.state

    this.props.addTask({ title, tagsString, issueNum, repo })
    this.setState({
      redirect: true
    })
  }

  render() {
    const {
      title,
      tags,
      issueNum,
      numDIDRequiredToAddTask,
      repo,
      redirect
    } = this.state

    if (redirect) return <Redirect to="/tasks" />

    return (
      <div>
        <Head title="Add Task" />
        <Grid>
          <Grid.Row>
            <PageTitling
              title="Propose"
              subtitle="DID holders can propose new work here"
            />
          </Grid.Row>
          <Grid.Row>
            <Grid.Column width={8}>
              <Form onSubmit={this.onSubmit}>
                <Form.Field>
                  <Input
                    onChange={this.onChangeIssueNum}
                    placeholder="Github Issue Number"
                    value={issueNum}
                  />
                </Form.Field>
                <Form.Field>
                  <Input
                    type="text"
                    placeholder="Title (<70 chars)"
                    onChange={this.onChangeTitle}
                    className=""
                    name="title"
                    value={title}
                  />
                </Form.Field>
                <Button size="large" color="green" type="submit">
                  Submit
                </Button>
              </Form>
            </Grid.Column>
            <Grid.Column width={8}>
              <Form>
                <Form.Field>
                  <Dropdown
                    fluid
                    multiple
                    onChange={this.onChangeTags}
                    options={tagsOptions}
                    placeholder="Tags"
                    search
                    selection
                    scrolling
                    value={tags}
                  />
                </Form.Field>
                <Form.Field>
                  <Dropdown
                    fluid
                    onChange={this.onChangeRepo}
                    options={[
                      {
                        key: 'ui',
                        value: 'ui',
                        text: 'UI'
                      },
                      {
                        key: 'contracts',
                        value: 'contracts',
                        text: 'Contracts'
                      }
                    ]}
                    placeholder="Repo"
                    search
                    selection
                    scrolling
                    value={repo}
                  />
                </Form.Field>
              </Form>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <Grid.Column width={16}>
              <Message>
                <Message.Header>Rules</Message.Header>
                <List bulleted>
                  <List.Item>
                    You must own at least {numDIDRequiredToAddTask} DID to
                    propose. This number changes according to the{' '}
                    <Link to="/parameters">proposalPctDIDToApprove</Link>{' '}
                    parameter
                  </List.Item>
                  <List.Item>
                    Your proposal can be anything, it doesn't necessarily have
                    to be <em>work</em>.
                  </List.Item>
                  <List.Item>
                    Remember that when you propose, it is likely that DID will
                    be issued for the completion of the task. The fewer DID the
                    better.
                  </List.Item>
                </List>
              </Message>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <Grid.Column width={8}>
              <Message>
                <Message.Header>Github Issue Number</Message.Header>
                <List bulleted>
                  <List.Item>
                    Create an issue in the appropriate Github repository to
                    determine the Github issue number
                  </List.Item>
                  <List.Item>
                    Your proposal can be anything, it doesn't necessarily have
                    to be <em>work</em>.
                  </List.Item>
                  <List.Item>
                    Discussion will happen on Github in the issue you create
                  </List.Item>
                </List>
              </Message>
            </Grid.Column>
            <Grid.Column width={8}>
              <Message>
                <Message.Header>Title</Message.Header>
                <List bulleted>
                  <List.Item>Enter the title of your task/proposal</List.Item>
                  <List.Item>
                    Limit the length of the title: the more characters the more
                    gas it will cost you
                  </List.Item>
                  <List.Item>
                    Example: 'Build this amazing new dapp that will change the
                    world'
                  </List.Item>
                  <List.Item>
                    People will like you more if you enter the same exact title
                    as your Github issue
                  </List.Item>
                </List>
              </Message>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <Grid.Column width={8}>
              <Message>
                <Message.Header>Tags</Message.Header>
                <List bulleted>
                  <List.Item>
                    Select the category of proposal you're making
                  </List.Item>
                  <List.Item>
                    Examples are: 'contracts', 'frontend', 'governance'
                  </List.Item>
                </List>
              </Message>
            </Grid.Column>
            <Grid.Column width={8}>
              <Message>
                <Message.Header>Repo</Message.Header>
                <List bulleted>
                  <List.Item>Choose the one that best fits</List.Item>
                </List>
              </Message>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  pendingTask: getPendingTask(state),
  numDIDRequiredToAddTask: getParameterValueByTitle(
    state,
    NUM_DID_REQUIRED_TO_ADD_TASK_PARAMETER_TITLE
  )
})

const mapDispatchToProps = dispatch => ({
  addTask: task => dispatch(addTask(task))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddTask)
