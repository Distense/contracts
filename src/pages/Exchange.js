import React, { Component } from 'react'

import { Grid, List, Message } from 'semantic-ui-react'

import ExchangeDIDForEther from './Exchange/ExchangeDIDForEther'
import ExchangeEtherForDID from './Exchange/ExchangeEtherForDID'

import Head from '../components/common/Head'
import PageTitling from '../components/common/PageTitling'

export default class Exchange extends Component {
  render() {
    return (
      <div>
        <Head title="Exchange DID for ETH" />
        <Grid>
          <PageTitling
            title="Exchange"
            subtitle="DID holders may exchange their DID into ether or vice versa"
          />
          <Grid.Column width={8}>
            <Message>
              <Message.Header>Distense Finances</Message.Header>
              <List bulleted>
                <List.Item>
                  ETH in Distense bank account: {this.props.numBankAccountEther}
                </List.Item>
                <List.Item>
                  Exchange rate: {this.props.didPerEtherExchangeRate} DID per
                  ETH
                </List.Item>
              </List>
            </Message>
          </Grid.Column>

          <Grid.Row style={{ paddingTop: '0px' }} columns={2}>
            <ExchangeDIDForEther {...this.props} />
            <ExchangeEtherForDID {...this.props} />
          </Grid.Row>
          <Grid.Row />
        </Grid>
      </div>
    )
  }
}
