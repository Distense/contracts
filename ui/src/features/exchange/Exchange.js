import React, { Component } from 'react'

import { Grid, List, Message } from 'semantic-ui-react'

import ExchangeDIDForEther from './components/ExchangeDIDForEther'
import ExchangeEtherForDID from './components/ExchangeEtherForDID'

import Head from '../../components/Head'
import PageTitling from '../../components/PageTitling'

export default class Exchange extends Component {
  render() {
    return (
      <div>
        <Head title="Exchange DID for ETH and vice versa" />
        <Grid>
          <PageTitling
            title="Exchange"
            subtitle="DID holders may exchange their DID into ether or vice versa"
          />
          <Grid.Column width={16}>
            <Message>
              <Message.Header>Distense Finances</Message.Header>
              <List bulleted>
                <List.Item>
                  ETH in Distense bank account: {this.props.numBankAccountEther}
                </List.Item>
                <List.Item>
                  Exchange rate: {this.props.didPerEtherExchangeRate} DID per
                  ETH
                </List.Item>
              </List>
            </Message>
          </Grid.Column>

          <Grid.Row style={{ paddingTop: '0px' }} columns={2}>
            <ExchangeDIDForEther {...this.props} />
            <ExchangeEtherForDID {...this.props} />
          </Grid.Row>
          <Grid.Row />
        </Grid>
      </div>
    )
  }
}
