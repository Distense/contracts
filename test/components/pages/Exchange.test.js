import React from 'react'
import { shallow } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { Exchange } from '../../../src/pages/Exchange'

function setup(numDIDOwned = null) {
  const props = {
    numDIDOwned,
    investEtherForDID: jest.fn(),
    exchangeDIDForEther: jest.fn()
  }
  let wrapper = shallow(<Exchange {...props} />)

  return { wrapper, props }
}

describe('<Exchange /> page component', () => {
  it('should contain some basic subcomponents', () => {
    const { wrapper } = setup()
    expect(wrapper.find('Button').length).toEqual(2)
    expect(wrapper.find('Grid').length).toEqual(1)
    expect(wrapper.find('Head').length).toEqual(0)
    expect(wrapper.find('GridRow').length).toEqual(2)
    expect(wrapper.find('Form').length).toEqual(2)
    expect(wrapper.find('FormField').length).toEqual(2)
    expect(wrapper.find('List').length).toEqual(2)
    expect(wrapper.find('Message').length).toEqual(2)
    expect(wrapper.find('MessageHeader').length).toEqual(2)
  })

  it('should set the initial state correctly', () => {
    const { wrapper } = setup()
    expect(wrapper.state('numDID')).toEqual('')
    expect(wrapper.state('numEther')).toEqual('')
    expect(wrapper.state('numDIDOwned')).toEqual(0)
  })

  it('should set numDidOwned to something other than 0 when comes through in props', () => {
    const { wrapper } = setup(2500)
    expect(wrapper.state('numDIDOwned')).toEqual(2500)
  })

  it('should contain two buttons (exchange DID for ether, invest ether for DID)', () => {
    const { wrapper } = setup()
    expect(
      wrapper
        .find('Button')
        .first()
        .children()
        .text()
    ).toEqual('Exchange DID for ether')
    expect(
      wrapper
        .find('Button')
        .at(1)
        .children()
        .text()
    ).toEqual('Invest ether for DID')
  })

  it('should set state when onSubmitExchangeDIDForEther is called', async () => {
    const { wrapper, props } = setup()
    const mockedEvent = { preventDefault: () => {} }
    await wrapper.instance().onSubmitExchangeDIDForEther(mockedEvent)
    expect(props.exchangeDIDForEther).toHaveBeenCalledWith({ numDID: '' })
  })

  it('should set state when onSubmitInvestEtherForDID is called', async () => {
    const { wrapper, props } = setup()
    const mockedEvent = { preventDefault: () => {} }
    await wrapper.instance().onSubmitInvestEtherForDID(mockedEvent)
    expect(props.investEtherForDID).toHaveBeenCalledWith({ numEther: '' })
  })

  it('should set state when onChangeNumEther is called', () => {
    const { wrapper } = setup()
    const mockedEvent = { target: { value: 25 } }
    wrapper.instance().onChangeNumEther(mockedEvent)
    expect(wrapper.state('numEther')).toEqual(25)
  })

  it('should set state when onChangeNumDID is called', () => {
    const { wrapper } = setup()
    const mockedEvent = { target: { value: 300 } }
    wrapper.instance().onChangeNumDID(mockedEvent)
    expect(wrapper.state('numDID')).toEqual(300)
  })

  test('snapshot', () => {
    const { props } = setup('executive')
    const mockstore = configureMockStore()
    const store = mockstore({})
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <Exchange {...props} />
          </MemoryRouter>
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
