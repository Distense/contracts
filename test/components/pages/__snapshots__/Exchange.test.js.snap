// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Exchange /> page component snapshot 1`] = `
<div>
  <div
    className="ui grid"
  >
    <div
      className="sixteen wide column"
    >
      <h2
        className="ui header"
      >
        Exchange
      </h2>
      <p
        style={
          Object {
            "fontSize": "1.3rem",
          }
        }
      >
        DID holders may exchange their DID into ether or vice versa
      </p>
    </div>
    <div
      className="two column row"
      style={
        Object {
          "paddingTop": "0px",
        }
      }
    >
      <div
        className="eight wide column"
      >
        <form
          action={undefined}
          className="ui form"
          onSubmit={[Function]}
        >
          <div
            className="field"
          >
            <div
              className="ui input"
            >
              <input
                disabled={undefined}
                onChange={[Function]}
                placeholder="receive ether, give up DID"
                tabIndex={undefined}
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="ui green large basic button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
            type="submit"
          >
            Exchange DID for ether
          </button>
        </form>
        <div
          className="ui message"
        >
          <div
            className="header"
          >
            Exchange DID For Ether
          </div>
          <div
            className="ui bulleted list"
            role="list"
          >
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You may exchange your DID for ether.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              With the current Ethereum coinbase account you are logged in with, you may exchange 
              executive
               DID
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              This option is available because not everyone can work solely for equity, unfortunately. We understand.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              The current exchange rate is 1000 DID per ether. This rate may be voted on in the
               
              <a
                href="/parameters"
                onClick={[Function]}
              >
                parameters
              </a>
               page
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You are relinquishing Distense voting rights by exchanging your DID for ether.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              This is a permanent, non-reversible, undoable, final action. There is no one who can undo this.
            </div>
          </div>
        </div>
      </div>
      <div
        className="eight wide column"
      >
        <form
          action={undefined}
          className="ui form"
          onSubmit={[Function]}
        >
          <div
            className="field"
          >
            <div
              className="ui input"
            >
              <input
                disabled={undefined}
                name="title"
                onChange={[Function]}
                placeholder="receive DID, give up ether"
                tabIndex={undefined}
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="ui green large basic button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
            type="submit"
          >
            Invest ether for DID
          </button>
        </form>
        <div
          className="ui message"
        >
          <div
            className="header"
          >
            Invest Ether for DID
          </div>
          <div
            className="ui bulleted list"
            role="list"
          >
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You may exchange ether only to the extent you have contributed to Distense. I.e, according to the number of DID you 
              <em>
                continue
              </em>
               to own
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              DID have a very high chance of being worthless. DID are not traded on any exchange and are non-transferable. Lots has to happen for DID to be worth anything.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              If you are idealistically aligned with Distense, have a super long-term timeframe in mind, then maybe, just maybe, you should consider giving up your precious ether for DID.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              This is a permanent, non-reversible, undoable, final action. There is no one who can undo this.
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    />
  </div>
</div>
`;
